src/celeritas/optical/WavelengthShiftParams.cc

```diff
@@ -0,0 +1,112 @@
+//---------------------------------------------------------------------------//
+#include "WavelengthShiftParams.hh"
+
+#include <vector>
+
+#include "corecel/data/CollectionBuilder.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/grid/GenericGridBuilder.hh"
+#include "celeritas/grid/GenericGridInserter.hh"
+#include "celeritas/io/ImportData.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct wavelength shift (WLS) data with imported data.
+ */
+std::shared_ptr<WavelengthShiftParams>
+WavelengthShiftParams::from_import(ImportData const& data)
+{
+    CELER_EXPECT(!data.optical_materials.empty());
+
+    if (!std::any_of(
+            data.optical_materials.begin(),
+            data.optical_materials.end(),
+            [](auto const& iter) { return static_cast<bool>(iter.wls); }))
+    {
+        // No wavelength shift data present
+        return nullptr;
+    }
+
+    Input input;
+    for (auto const& mat : data.optical_materials)
+    {
+        input.data.push_back(mat.wls);
+    }
+    return std::make_shared<WavelengthShiftParams>(std::move(input));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with wavelength shift (WLS) input data.
+ */
+WavelengthShiftParams::WavelengthShiftParams(Input const& input)
+{
+    CELER_EXPECT(input.data.size() > 0);
+    HostVal<WavelengthShiftData> data;
+
+    CollectionBuilder wls_scalars{&data.wls_scalars};
+    GenericGridInserter insert_energy_cdf(&data.reals, &data.energy_cdf);
+    for (auto const& wls : input.data)
+    {
+        // Store WLS component tabulated as a function of photon energy
+        auto const& comp_vec = wls.component;
+
+        if (comp_vec.x.empty())
+        {
+            // No WLS data for this material
+            continue;
+        }
+
+        CELER_VALIDATE(wls.mean_num_photons > 0,
+                       << "invalid mean_num_photons=" << wls.mean_num_photons
+                       << " (should be positive)");
+        CELER_VALIDATE(wls.time_constant > 0,
+                       << "invalid time_constant=" << wls.time_constant
+                       << " (should be positive)");
+
+        // Scalar properties
+        ScalarPropertyWLS scalars;
+        scalars.mean_num_photons = wls.mean_num_photons;
+        scalars.time_constant = wls.time_constant;
+        wls_scalars.push_back(scalars);
+
+        // Calculate the WLS cumulative probability of the emission spectrum
+        std::vector<real_type> cdf(comp_vec.x.size());
+
+        CELER_ASSERT(comp_vec.y[0] > 0);
+        // The value of cdf at the low edge is zero by default
+        cdf[0] = 0;
+        for (size_type i = 1; i < comp_vec.x.size(); ++i)
+        {
+            cdf[i] = cdf[i - 1]
+                     + real_type(0.5) * (comp_vec.x[i] - comp_vec.x[i - 1])
+                           * (comp_vec.y[i] + comp_vec.y[i - 1]);
```

Comment:
A little function `real_type integrate_trapezoid(Real2 lo, Real2 hi)` would improve clarity here.

Reply:
Can you help this when you add the CDF sampler class too?  Thanks.

Reply:
I'll leave a note as a to-do.